lib/aescrypt.rb

```diff
@@ -28,19 +28,27 @@
 
 require 'openssl'
 require 'base64'
+require 'securerandom'
+
+require_relative 'migrator'
 
 module AESCrypt
-  def self.encrypt(message, password)
-    Base64.encode64(self.encrypt_data(message.to_s.strip, self.key_digest(password), nil, "AES-256-CBC"))
+  def self.encrypt(message, password, salt = nil, iv = nil)
+    iv ||= SecureRandom.bytes 16
+    salt, key = self.key_digest(password, salt)
+    return salt, iv, Base64.encode64(self.encrypt_data(message.to_s.strip, key, iv, 'AES-256-CBC'))
   end
 
-  def self.decrypt(message, password)
+  def self.decrypt(message, password, salt, iv)
     base64_decoded = Base64.decode64(message.to_s.strip)
-    self.decrypt_data(base64_decoded, self.key_digest(password), nil, "AES-256-CBC")
+    slt, key = self.key_digest(password, salt)
+    self.decrypt_data(base64_decoded, key, iv, 'AES-256-CBC')
   end
 
-  def self.key_digest(password)
-    OpenSSL::Digest::SHA256.new(password).digest
+  def self.key_digest(password, salt = nil)
+    salt ||= SecureRandom.bytes 32
+    digest = OpenSSL::Digest::SHA256.new
+    return salt, OpenSSL::PKCS5.pbkdf2_hmac(password, salt, 10000, digest.digest_length, digest)
```

Comment:
An HMAC is a message authentication code. You probably took this from [the PKCS5 example](http://ruby-doc.org/stdlib-2.0.0/libdoc/openssl/rdoc/OpenSSL/PKCS5.html).

It's not that what you're doing is wrong, exactly, but I'm not sure it is what you need. The purposes of an HMAC is that given a shared secret (here: the salt), you can authenticate that a message (here: the password) hasn't been tampered with.

You would normally send the message and the digest across an insecure medium, relying on the HMAC (and the shared secret that's been communicated securely beforehand) to verify the message has not been tampered with.

The example also suggests to generate cipher keys this way (that's questionable), and to store passwords that way.

Storing passwords in the form of salt and HMAC works just fine. If that's your intended use, go for it.

Generating cipher keys with an HMAC is simply unnecessary, and forces you to also deal with the salt.